icu/icu4c/source/common/uprefs.cpp

```diff
@@ -0,0 +1,553 @@
+    RETURN_WITH_ALLOCATION_ERROR_IF(U_FAILURE(*status), status);
+    
+    int32_t result = GetLocaleInfoExWrapper(LOCALE_NAME_USER_DEFAULT, LOCALE_SNAME, NLSLocale.getAlias(), neededBufferSize, status);
+
+    RETURN_VALUE_IF(U_FAILURE(*status), CharString());
+    
+    if (getSorting) //We determine if we want the locale (for example, en-US) or the sorting method (for example, phonebk)
+    {
+        // We use LOCALE_SNAME to get the sorting method (if any). So we need to keep
+        // only the sorting bit after the _, removing the locale name.
+        // Example: from "de-DE_phoneb" we only want "phoneb"
+        const wchar_t * startPosition = wcschr(NLSLocale.getAlias(), L'_');
+
+        // Note: not finding a "_" is not an error, it means the user has not selected an alternate sorting method, which is fine.
+        if (startPosition != nullptr) 
+        {
+            CharString sortingSystem = getSortingSystemBCP47FromNLSType(startPosition + 1, status);
+
+            if (sortingSystem.length() == 0)
+            {
+                *status = U_UNSUPPORTED_ERROR;
+                return CharString();
+            }
+            return sortingSystem;
+        }
+    }
+    else
+    {
+        // The NLS locale may include a non-default sort, such as de-DE_phoneb. We only want the locale name before the _.
+        wchar_t * position = wcschr(NLSLocale.getAlias(), L'_');
+        if (position != nullptr)
+        {
+            *position = L'\0';
+        }
+
+        CharString languageTag;
+        int32_t resultCapacity = 0;
+        languageTag.getAppendBuffer(neededBufferSize, neededBufferSize, resultCapacity, *status);
+        RETURN_WITH_ALLOCATION_ERROR_IF(U_FAILURE(*status), status);
+
+        int32_t unitsWritten = 0;
+        u_strToUTF8(languageTag.data(), neededBufferSize, &unitsWritten, reinterpret_cast<UChar*>(NLSLocale.getAlias()), neededBufferSize, status);
+        RETURN_VALUE_IF(U_FAILURE(*status), CharString());
+        return languageTag;
+    }
+
+    return CharString();
+}
+
+CharString getCalendarSystem_impl(UErrorCode* status)
+{
+    int32_t NLSCalendar = 0;
+
+    GetLocaleInfoExWrapper(LOCALE_NAME_USER_DEFAULT, LOCALE_ICALENDARTYPE | LOCALE_RETURN_NUMBER, reinterpret_cast<PWSTR>(&NLSCalendar), sizeof(NLSCalendar) / sizeof(wchar_t), status);
+
+    RETURN_VALUE_IF(U_FAILURE(*status), CharString());
+
+    CharString calendar(getCalendarBCP47FromNLSType(NLSCalendar, status), *status);
+    RETURN_FAILURE_STRING_WITH_STATUS_IF(calendar.length() == 0, U_UNSUPPORTED_ERROR, status);
+
+    return calendar;
+}
+
+CharString getCurrencyCode_impl(UErrorCode* status)
+{
+    int32_t neededBufferSize = GetLocaleInfoExWrapper(LOCALE_NAME_USER_DEFAULT, LOCALE_SINTLSYMBOL, nullptr, 0, status);
+    
+    RETURN_VALUE_IF(U_FAILURE(*status), CharString());
+    
+    MaybeStackArray<wchar_t, 40> NLScurrencyData(neededBufferSize, *status);
```

Comment:
We could probably reduce this a bit. :)
Looking at the [docs](https://docs.microsoft.com/en-us/windows/win32/intl/locale-sintlsymbol), it says the max length is 9, so 10 would probably cover it. 

Probably add a comment like:
`// Per https://docs.microsoft.com/en-us/windows/win32/intl/locale-sintlsymbol the max length is 9.`
```suggestion
    MaybeStackArray<wchar_t, 10> NLScurrencyData(neededBufferSize, *status);
```